src/sage/groups/finitely_presented.py

```diff
@@ -361,6 +361,57 @@ def __call__(self, *values, **kwds):
                     raise ValueError('the values do not satisfy all relations of the group')
         return super().__call__(values)
 
+    def _normal_form_by_gap_nffunction(self) -> GapElement:
+        """
+        Internal method that calls ``FpElementNFFunction`` to compute some normal form
+        for this group element.
+        Note that this is not guaranteed to terminate.
+
+        .. SEEALSO::
+
+            Another way to compute a normal form is to use
+            :meth:`FinitelyPresentedGroup.confluent_rewriting_system`.
+
+        OUTPUT:
+
+        A :class:`~GapElement` such that if ``x == y``
+        then ``x._normal_form_by_gap_nffunction()`` is
+        determined only by the value of this group element,
+        and not the Tietze list.
+
+        TESTS::
+
+            sage: F.<a,b> = FreeGroup()
+            sage: G = F / [a*b/a/b]
+            sage: G(a*b) == G(b*a)
+            True
+            sage: G(a*b).Tietze() == G(b*a).Tietze()
+            False
+            sage: G(a*b)._normal_form_by_gap_nffunction() == G(b*a)._normal_form_by_gap_nffunction()
+            True
+        """
+        x = self.gap()
+        return x.FamilyObj().FpElementNFFunction()(x.UnderlyingElement())
+
+    def __hash__(self) -> int:
+        """
+        Return some hash value, such that if ``x == y``
+        then ``hash(x) == hash(y)``.
+        Note that this is not guaranteed to terminate.
```

Comment:
```suggestion
    def __hash__(self) -> int:
        """
        Return the hash of ``self`` based on a normal form.

        .. WARNING::

            Note that this is not guaranteed to terminate.
```
I am still a bit worried about this taking a long time and/or failing. Additionally, we should cache the result.

Reply:
I think GAP already cache the result (surely it must)... well, it's actually hard to find a group where this terminates in a few seconds to test this guess.